### ISSUE TITLE
ansible relocatable bottle

### ISSUE THREAD

**Issue by lucafavatella** [Homebrew/homebrew-core#9424]
# General troubleshooting steps

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [ ] Confirmed this is problem with specific formulae and not Homebrew/brew? If it's a general Homebrew/brew problem please file this issue at https://github.com/Homebrew/brew/issues/new
  - I have the impression this may be a symptom of bottle non-relocatability for all formulas using python virtualenv but I am not sure so I file this formula-specific ticket first.

## Bug reports

Brief summary of my issue:

The bottle of the [ansible formula](https://github.com/Homebrew/homebrew-core/blob/f518350a531002634d80d921d63d6ec136ebaf90/Formula/ansible.rb) is not relocatable, that is a regression from [a previous version](https://github.com/Homebrew/brew/pull/382#issuecomment-227308942).

### Output of `brew config`

```
HOMEBREW_VERSION: 1.1.8-64-gc50722268
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: c5072226871196d2eb3265d930daff87968d23e8
Last commit: 26 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 61869e31ad68eccf6bd60d312255bb4e2ff59efd
Core tap last commit: 6 hours ago
HOMEBREW_PREFIX: /Users/luca/homebrew
HOMEBREW_REPOSITORY: /Users/luca/homebrew
HOMEBREW_CELLAR: /Users/luca/homebrew/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: dual-core 64-bit penryn
Homebrew Ruby: 2.0.0-p481
Clang: 7.0 build 700
Git: 2.11.0 => /Users/luca/homebrew/bin/git
Perl: /usr/bin/perl
Python: /Users/luca/homebrew/bin/python => /Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_112, 1.8.0_102, 1.8.0_92, 1.8.0_77, 1.8.0_74, 1.8.0_66
macOS: 10.10.5-x86_64
Xcode: N/A
CLT: 7.2.0.0.1.1447826929
X11: N/A
```

### Output of `brew doctor`

```
Warning: The filesystem on / appears to be case-sensitive.
The default macOS filesystem is case-insensitive. Please report any apparent problems.

Warning: Your Homebrew's prefix is not /usr/local.
You can install Homebrew anywhere you want but some bottles (binary packages)
can only be used with a /usr/local prefix and some formulae (packages)
may not build correctly with a non-/usr/local prefix.

Warning: Some installed formula are missing dependencies.
You should `brew install` the missing dependencies:
  brew install doxygen

Run `brew missing` for more details.
```

Output of `brew missing`:
```
libevent: doxygen
tmux: doxygen
```

### Output of `brew gist-logs ansible`

https://gist.github.com/0171eb75017283f1f65d5727552bc306

### Details of my issue

I ran `brew install --build-bottle ansible` and `brew bottle --verbose ansible`. I paste the relevant excerpts below.

```
==> Detecting if ansible-2.2.1.0.yosemite.bottle.1.tar.gz is relocatable...
Warning: String '/Users/luca/homebrew' still exists in these files:
```

The following lines are symptoms unknown to me:
``` 
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/config/Makefile
 --> match '/Users/luca/homebrew/opt/python/Frameworks/Python.framework/Versions/2.7' at offset 0xd75
 --> match 'datarootdir=    /Users/luca/homebrew/opt/python/share' at offset 0xe29
 --> match '/Users/luca/homebrew/opt/python/Frameworks' at offset 0x1205
 --> match 'PYTHONFRAMEWORKINSTALLDIR= /Users/luca/homebrew/opt/python/Frameworks/Python.framework' at offset 0x1230
 --> match ''--prefix=/Users/luca/homebrew/opt/python' '--enable-ipv6' '--datarootdir=/Users/luca/homebrew/opt/python/share' '--datadir=/Users/luca/homebrew/opt/python/share' '--enable-framework=/Users/luca/homebrew/opt/python/Frameworks' '--without-ensurepip' '--without-gcc' 'MACOSX_DEPLOYMENT_TARGET=10.10' 'CC=clang' 'PKG_CONFIG_PATH=/Users/luca/homebrew/opt/sqlite/lib/pkgconfig:/Users/luca/homebrew/opt/openssl/lib/pkgconfig' 'PKG_CONFIG_LIBDIR=/usr/lib/pkgconfig:/Users/luca/homebrew/Library/Homebrew/os/mac/pkgconfig/10.10'' at offset 0x1571
```

The following lines (`*.pyc` and `*.pyo`) are supposed to have [been fixed](https://github.com/Homebrew/brew/pull/382) but maybe they came back for [virtualenv](https://github.com/Homebrew/homebrew-core/blob/f518350a531002634d80d921d63d6ec136ebaf90/Formula/ansible.rb#L2):
```
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/__init__.pyc
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyR' at offset 0x683
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x9e0
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x1010
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x11a6
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/__init__.pyo
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyR' at offset 0x683
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x9e0
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x1010
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/__init__.pyt' at offset 0x11a6
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/aliases.pyc
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/aliases.pyt' at offset 0x1f9b
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/aliases.pyo
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/aliases.pyt' at offset 0x1f9b
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/ascii.pyc
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/ascii.pyR' at offset 0x221
[...]
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/ascii.pyo
 --> match '/Users/luca/homebrew/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings/ascii.pyR' at offset 0x221
[...]
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/zlib_codec.pyc
[...]
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings/zlib_codec.pyo
[...]
```

These lines appear related to python and openssl links:
```
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/lib-dynload/_hashlib.so
 --> links to /Users/luca/homebrew/opt/openssl/lib/libssl.1.0.0.dylib
 --> links to /Users/luca/homebrew/opt/openssl/lib/libcrypto.1.0.0.dylib
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/lib-dynload/_sqlite3.so
 --> links to /Users/luca/homebrew/opt/sqlite/lib/libsqlite3.0.dylib
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/lib-dynload/_ssl.so
 --> links to /Users/luca/homebrew/opt/openssl/lib/libssl.1.0.0.dylib
 --> links to /Users/luca/homebrew/opt/openssl/lib/libcrypto.1.0.0.dylib
/Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/lib-dynload/readline.so
 --> links to /Users/luca/homebrew/opt/readline/lib/libreadline.7.dylib
Warning: Absolute symlink starting with /Users/luca/homebrew:
  /Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/include/python2.7 -> /Users/luca/homebrew/opt/python/Frameworks/Python.framework/Versions/2.7/include/python2.7
  /Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/config -> /Users/luca/homebrew/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config
  /Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/encodings -> /Users/luca/homebrew/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings
  /Users/luca/homebrew/Cellar/ansible/2.2.1.0/libexec/lib/python2.7/lib-dynload -> /Users/luca/homebrew/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload
```

**Comment by MikeMcQuaid** [comment]
> The bottle of the ansible formula is not relocatable, that is a regression from a previous version.

Note we don't generally consider these to be a regression.


> The following lines (*.pyc and *.pyo) are supposed to have been fixed but maybe they came back for virtualenv

This does look like a possible regression so leaving this open until @tdsmith has some thoughts.

**Comment by ilovezfs** [comment]
Ping @tdsmith

**Comment by MikeMcQuaid** [comment]
@tdsmith any thoughts here?

**Comment by MikeMcQuaid** [comment]
@tdsmith Gentle nudge on this again.

**Comment by tdsmith** [comment]
The relevant bottle output: https://gist.github.com/83a59f9e062706913f06a075aad38fbc

Looks like the cleaner does need some help, yeah.

**Comment by tdsmith** [comment]
Oh, which was posted at the top of this thread, heh. Anyway, I'll take a look.

**Comment by tdsmith** [comment]
Ah; these .pyc files are not actually in the keg. There's some symlink following the relocator hasn't figured out.

```
tim@rocketman:~$ realpath /usr/local/Cellar/ansible/2.2.2.0/libexec/lib/python2.7/encodings
/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings
tim@rocketman:~$ ls -l /usr/local/Cellar/ansible/2.2.2.0/libexec/lib/python2.7/encodings
lrwxr-xr-x  1 tim  admin  86 Mar 27 13:33 /usr/local/Cellar/ansible/2.2.2.0/libexec/lib/python2.7/encodings -> /usr/local/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/encodings
```

**Comment by tdsmith** [comment]
The symlinks are being handled correctly but `bottle` is complaining about files it shouldn't be seeing. I'm a little confused because the files to scan are enumerated with `Utils.popen_read("/usr/bin/fgrep", "-lr", string, to_s)`. The man page says that by default it shouldn't recurse into symlinked directories, and yet.

**Comment by tdsmith** [comment]
Heh. The man page is wrong! Adding `-p` or `-O` to that invocation correctly suppresses the bogus output, "This is the default" be damned. Unfortunately, those not recognized by GNU grep.

**Comment by tdsmith** [comment]
This should be fixed by #2442.

**Comment by lucafavatella** [comment]
@tdsmith It seems fixed for me. Thanks!
